allennlp/training/optimizers.py

```diff
@@ -31,143 +31,332 @@
-            # Those will be included in the last entry of parameter_groups.
-            parameter_groups: Any = [{"params": []} for _ in range(len(groups) + 1)]
-            # add the group specific kwargs
-            for k in range(len(groups)):
-                parameter_groups[k].update(groups[k][1].as_dict())
-
-            regex_use_counts: Dict[str, int] = {}
-            parameter_group_names: List[set] = [set() for _ in range(len(groups) + 1)]
-            for name, param in model_parameters:
-                # Determine the group for this parameter.
-                group_index = None
-                for k, group_regexes in enumerate(groups):
-                    for regex in group_regexes[0]:
-                        if regex not in regex_use_counts:
-                            regex_use_counts[regex] = 0
-                        if re.search(regex, name):
-                            if group_index is not None and group_index != k:
-                                raise ValueError(
-                                    "{} was specified in two separate parameter groups".format(name)
-                                )
-                            group_index = k
-                            regex_use_counts[regex] += 1
-
-                if group_index is not None:
-                    parameter_groups[group_index]["params"].append(param)
-                    parameter_group_names[group_index].add(name)
-                else:
-                    # the default group
-                    parameter_groups[-1]["params"].append(param)
-                    parameter_group_names[-1].add(name)
-
-            # log the parameter groups
-            logger.info("Done constructing parameter groups.")
-            for k in range(len(groups) + 1):
-                group_options = {
-                    key: val for key, val in parameter_groups[k].items() if key != "params"
-                }
-                logger.info("Group %s: %s, %s", k, list(parameter_group_names[k]), group_options)
-            # check for unused regex
-            for regex, count in regex_use_counts.items():
-                if count == 0:
-                    logger.warning(
-                        "When constructing parameter groups, %s does not match any parameter name",
-                        regex,
-                    )
-
-        else:
-            parameter_groups = [param for name, param in model_parameters]
-
-        # Log the number of parameters to optimize
-        num_parameters = 0
-        for parameter_group in parameter_groups:
-            if isinstance(parameter_group, dict):
-                num_parameters += sum(parameter.numel() for parameter in parameter_group["params"])
-            else:
-                num_parameters += parameter_group.numel()
-        logger.info("Number of trainable parameters: %s", num_parameters)
-
-        # By default we cast things that e.g. look like floats to floats before handing them
```

Comment:
So if I understand correctly, this just always happens now? I don't see any usages of it in the negative (even with `git log -S infer_type_and_cast -p`), so it's unclear to me why the option existed in the first place. Guessing that was your thinking too... :)

Reply:
Basically, that was added because of craziness with using `std.extvar` (or something) in jsonnet / environment variables with json.  Now that we have explicit constructors, it is entirely unnecessary.